ntp-proto/src/peer.rs

```diff
@@ -228,29 +226,25 @@ impl Peer {
         send_time: NtpTimestamp,
         recv_time: NtpTimestamp,
     ) -> Result<PeerSnapshot, IgnoreReason> {
-        // we're expecting a packet
-        debug_assert!(self.next_expected_origin.is_some());
-
-        // the transmit_timestamp field was not changed from the bogus value we put into it
-        let transmit_unchanged = Some(message.transmit_timestamp) == self.next_expected_origin;
-
-        // the origin_timestamp changed; the server is not allowed to modify this field
-        let origin_changed = Some(message.origin_timestamp) != self.next_expected_origin;
-
         if message.mode != NtpAssociationMode::Server {
             // we currently only support a client <-> server association
             Err(IgnoreReason::InvalidMode)
-        } else if transmit_unchanged || origin_changed {
-            Err(IgnoreReason::InvalidPacketTime)
         } else if message.is_kiss_rate() {
+            // KISS packets may not have correct timestamps at all, handle them anyway
             self.remote_min_poll_interval =
                 Ord::max(self.remote_min_poll_interval.inc(), self.last_poll_interval);
             Err(IgnoreReason::KissIgnore)
         } else if message.is_kiss_rstr() || message.is_kiss_deny() {
+            // KISS packets may not have correct timestamps at all, handle them anyway
             Err(IgnoreReason::KissDemobilize)
         } else if message.is_kiss() {
             // Ignore unrecognized control messages
             Err(IgnoreReason::KissIgnore)
+        } else if Some(message.origin_timestamp) != self.next_expected_origin {
+            // Packets should be a response to a previous request from us,
+            // if not just ignore. Note that this might also happen when
+            // we reset between sending the request and receiving the response.s
+            Err(IgnoreReason::InvalidPacketTime)
```

Comment:
I haven't seen that requirement in the spec, which page was it on? We don't explicitly check it here, and I guess it is probably more important for servers when handling incoming requests, as packet duplication in the network could give some weird effects otherwise.

Reply:
the relevant test is

```
   | 2 bogus packet           |                                        |
   | 3 invalid                | One or more timestamp fields are       |
   |                          | invalid. This normally happens in      |
   |                          | symmetric modes when one peer sends    |
   |                          | the first packet to the other and      |
   |                          | before the other has received its      |
   |                          | first reply. 
```

so, not very explicit.

Reply:
I don't think a repeated transmit should necessarily make a packet invalid for us. Although very unlikely, it might actually happen around things like leap seconds.